File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_LOGONRESPONSE2.cs
﻿using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_LOGONRESPONSE2 : Message
    {
        protected enum Statuses : UInt32
        {
            Success = 0,
            AccountNotFound = 1,
            BadPassword = 2,
            AccountClosed = 6,
        };

        public SID_LOGONRESPONSE2()
        {
            Id = (byte)MessageIds.SID_LOGONRESPONSE2;
            Buffer = new byte[0];
        }

        public SID_LOGONRESPONSE2(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_LOGONRESPONSE2;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] SID_LOGONRESPONSE2 ({4 + Buffer.Length} bytes)");

                        if (Buffer.Length < 29)
                            throw new GameProtocolViolationException(context.Client, "SID_LOGONRESPONSE2 buffer must be at least 29 bytes");

                        /**
                         * (UINT32)     Client Token
                         * (UINT32)     Server Token
                         *  (UINT8)[20] Password Hash
                         * (STRING)     Username
                         */

                        using var m = new MemoryStream(Buffer);
                        using var r = new BinaryReader(m);

                        var clientToken = r.ReadUInt32();
                        var serverToken = r.ReadUInt32();
                        var passwordHash = r.ReadBytes(20);
                        context.Client.GameState.Username = r.ReadString();

                        Battlenet.Common.AccountsDb.TryGetValue(context.Client.GameState.Username, out Account account);

                        if (account == null)
                        {
                            return new SID_LOGONRESPONSE2().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {{ "status", Statuses.AccountNotFound }}));
                        }

                        var passwordHashDb = (byte[])account.Get(Account.PasswordKey, new byte[20]);
                        var compareHash = OldAuth.CheckDoubleHashData(passwordHashDb, clientToken, serverToken);
                        if (!compareHash.SequenceEqual(passwordHash))
                        {
                            account.Set(Account.FailedLogonsKey, ((UInt32)account.Get(Account.FailedLogonsKey, (UInt32)0)) + 1);
                            return new SID_LOGONRESPONSE2().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {{ "status", Statuses.BadPassword }}));
                        }

                        var flags = (Account.Flags)account.Get(Account.FlagsKey, Account.Flags.None);
                        if ((flags & Account.Flags.Closed) != 0)
                        {
                            account.Set(Account.FailedLogonsKey, ((UInt32)account.Get(Account.FailedLogonsKey, (UInt32)0)) + 1);
                            return new SID_LOGONRESPONSE2().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {{ "status", Statuses.AccountClosed }}));
                        }
                        
                        context.Client.GameState.ActiveAccount = account;
                        context.Client.GameState.FailedLogons = (UInt32)account.Get(Account.FailedLogonsKey, (UInt32)0);
                        context.Client.GameState.LastLogon = (DateTime)account.Get(Account.LastLogonKey, DateTime.Now);

                        account.Set(Account.FailedLogonsKey, (UInt32)0);
                        account.Set(Account.IPAddressKey, context.Client.RemoteEndPoint.ToString().Split(":")[0]);
                        account.Set(Account.LastLogonKey, DateTime.Now);
                        account.Set(Account.PortKey, context.Client.RemoteEndPoint.ToString().Split(":")[1]);

                        lock (Battlenet.Common.ActiveAccounts)
                        {
                            var serial = 1;
                            var onlineName = context.Client.GameState.Username;

                            while (Battlenet.Common.ActiveAccounts.ContainsKey(onlineName))
                            {
                                onlineName = $"{context.Client.GameState.Username}#{++serial}";
                            }

                            context.Client.GameState.OnlineName = onlineName;
                            Battlenet.Common.ActiveAccounts.Add(onlineName, account);
                        }

                        context.Client.GameState.Username = (string)account.Get(Account.UsernameKey, context.Client.GameState.Username);

                        lock (Battlenet.Common.ActiveGameStates)
                        {
                            Battlenet.Common.ActiveGameStates.Add(context.Client.GameState.OnlineName, context.Client.GameState);
                        }

                        return new SID_LOGONRESPONSE2().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient, new Dictionary<string, object> {{ "status", Statuses.Success }}));
                    }
                case MessageDirection.ServerToClient:
                    {
                        /**
                         * (UINT32) Status
                         * (STRING) Additional information (optional)
                         */

                        var info = (byte[])(context.Arguments.ContainsKey("info") ? context.Arguments["info"] : new byte[0]);

                        Buffer = new byte[4 + info.Length];

                        using var m = new MemoryStream(Buffer);
                        using var w = new BinaryWriter(m);

                        w.Write((UInt32)(Statuses)context.Arguments["status"]);
                        w.Write(info);
                        if (info.Length > 0) w.Write((byte)0);

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] SID_LOGONRESPONSE2 ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_LOGONRESPONSE2.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_LOGONRESPONSE2
        Inherits Message

        Protected Enum Statuses As UInt32
            Success = 0
            AccountNotFound = 1
            BadPassword = 2
            AccountClosed = 6
        End Enum

        Public Sub New()
            Id = CByte(MessageIds.SID_LOGONRESPONSE2)
            Buffer = Array.Empty(Of Byte)()
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_LOGONRESPONSE2)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal varContext As MessageContext) As Boolean
            Dim rereAccount As Account = Nothing

            Select Case varContext.Direction
                Case MessageDirection.ClientToServer
                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, varContext.Client.RemoteEndPoint, $"[{Common.DirectionToString(varContext.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    If Buffer.Length < 29 Then Throw New GameProtocolViolationException(varContext.Client, $"{MessageName(Id)} buffer must be at least 29 bytes")

                    Dim clientToken, serverToken As UInt32
                    Dim passwordHash() As Byte
                    Using m = New MemoryStream(Buffer)
                        Using r = New BinaryReader(m)
                            clientToken = r.ReadUInt32()
                            serverToken = r.ReadUInt32()
                            passwordHash = r.ReadBytes(20)
                            varContext.Client.GameState.Username = r.ReadString()
                        End Using
                    End Using

                    Battlenet.Common.AccountsDb.TryGetValue(varContext.Client.GameState.Username, rereAccount)

                    If rereAccount Is Nothing Then
                        Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, varContext.Client.RemoteEndPoint, $"Account [{varContext.Client.GameState.Username}] does not exist")
                        Return New SID_LOGONRESPONSE2().Invoke(New MessageContext(varContext.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object) From {
                            {"status", Statuses.AccountNotFound}
                        }))
                    End If

                    Dim passwordHashDb = CType(rereAccount.[Get](Account.PasswordKey, New Byte(19) {},), Byte())
                    Dim compareHash = OldAuth.CheckDoubleHashData(passwordHashDb, clientToken, serverToken)

                    If Not compareHash.SequenceEqual(passwordHash) Then
                        Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, varContext.Client.RemoteEndPoint, $"Account [{varContext.Client.GameState.Username}] logon failed password mismatch")
                        rereAccount.[Set](Account.FailedLogonsKey, (CUInt(rereAccount.[Get](Account.FailedLogonsKey, CUInt(0),))) + 1)
                        Return New SID_LOGONRESPONSE2().Invoke(New MessageContext(varContext.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object) From {
                            {"status", Statuses.BadPassword}
                        }))
                    End If

                    Dim flags As Account.Flags = CType(rereAccount.[Get](Account.FlagsKey, Account.Flags.None,), Account.Flags)

                    If (flags And Account.Flags.Closed) <> 0 Then
                        Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, varContext.Client.RemoteEndPoint, $"Account [{varContext.Client.GameState.Username}] logon failed account closed")
                        rereAccount.[Set](Account.FailedLogonsKey, (CUInt(rereAccount.[Get](Account.FailedLogonsKey, CUInt(0),))) + 1)
                        Return New SID_LOGONRESPONSE2().Invoke(New MessageContext(varContext.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object) From {
                            {"status", Statuses.AccountClosed}
                        }))
                    End If

                    varContext.Client.GameState.ActiveAccount = rereAccount
                    varContext.Client.GameState.FailedLogons = CUInt(rereAccount.[Get](Account.FailedLogonsKey, CUInt(0),))
                    varContext.Client.GameState.LastLogon = CDate(rereAccount.[Get](Account.LastLogonKey, DateTime.Now,))
                    rereAccount.[Set](Account.FailedLogonsKey, CUInt(0))
                    rereAccount.[Set](Account.IPAddressKey, varContext.Client.RemoteEndPoint.ToString().Split(":")(0))
                    rereAccount.[Set](Account.LastLogonKey, DateTime.Now)
                    rereAccount.[Set](Account.PortKey, varContext.Client.RemoteEndPoint.ToString().Split(":")(1))

                    SyncLock Battlenet.Common.ActiveAccounts
                        Dim serial = 1
                        Dim onlineName = varContext.Client.GameState.Username

                        While Battlenet.Common.ActiveAccounts.ContainsKey(onlineName.ToLower())
                            onlineName = $"{varContext.Client.GameState.Username}#{System.Threading.Interlocked.Increment(serial)}"
                        End While

                        varContext.Client.GameState.OnlineName = onlineName
                        Battlenet.Common.ActiveAccounts.Add(onlineName.ToLower(), rereAccount)
                    End SyncLock

                    varContext.Client.GameState.Username = CStr(rereAccount.[Get](Account.UsernameKey, varContext.Client.GameState.Username))

                    SyncLock Battlenet.Common.ActiveGameStates
                        Battlenet.Common.ActiveGameStates.Add(varContext.Client.GameState.OnlineName, varContext.Client.GameState)
                    End SyncLock

                    Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Client_Game, varContext.Client.RemoteEndPoint, $"Account [{varContext.Client.GameState.Username}] logon success as [{varContext.Client.GameState.OnlineName}]")
                    Return New SID_LOGONRESPONSE2().Invoke(New MessageContext(varContext.Client, MessageDirection.ServerToClient, New Dictionary(Of String, Object) From {
                        {"status", Statuses.Success}
                    }))
                Case MessageDirection.ServerToClient
                    Dim status = CUInt(CType(varContext.Arguments("status"), Statuses))
                    Dim info = CType((If(varContext.Arguments.ContainsKey("info"), varContext.Arguments("info"), Array.Empty(Of Byte)())), Byte())
                    Buffer = New Byte(4 + info.Length - 1) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(status)
                            w.Write(info)
                            If info.Length > 0 Then w.Write(CByte(0))
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, varContext.Client.RemoteEndPoint, $"[{Common.DirectionToString(varContext.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes) (status: 0x{status})")
                    varContext.Client.Send(ToByteArray(varContext.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
